### PR TITLE
Wrapper for the Mixed Integer Linear Minimization Oracle

### PR DESCRIPTION
**Pull request by dhendryc**
Similar to the Linear Minimization Oracle in `FrankWolfe.jl`, the aim is to also support solvers that do not implement the `MathOptInterface` package methods.

### CODE REVIEW

File: src/heuristics.jl [hunk 1]
```diff
@@ -1,3 +1,8 @@
+"""
```
Comment (matbesancon) [inline src/heuristics.jl:1]: no need for a docstring at the file level, it should usually be clear from the name of the file. For the TODO, better have it as an issue

File: src/interface.jl [hunk 2]
```diff
@@ -3,7 +3,7 @@
    
 f                      - objective function oracle. 
 g                      - oracle for the gradient of the objective. 
-lmo                    - a MIP solver instance (SCIP) encoding the feasible region.    
+blmo                    - a MIP solver instance (e.g.SCIP) encoding the feasible region. Has to be of type BoundedLinearMinimizationOracle (see lmo_wrapper).
```
Comment (matbesancon) [inline src/interface.jl:6]:
```suggestion
blmo                    - a MIP solver instance (e.g., SCIP) encoding the feasible region. Has to be of type `BoundedLinearMinimizationOracle` (see `lmo_wrapper.jl`).    
```

File: src/Boscia.jl [hunk 3]
```diff
@@ -3,6 +3,8 @@ module Boscia
 using FrankWolfe
 import FrankWolfe: compute_extreme_point
 export compute_extreme_point
+import Base: copy
+export copy
```
Comment (matbesancon) [inline src/Boscia.jl:7]: why do we need to reexport copy?

File: src/MOI_bounded_oracle.jl [hunk 4]
```diff
@@ -75,13 +75,13 @@ end
 Get the list of lower bounds.
 """
 function get_lower_bound_list(blmo::MathOptBLMO) 
-    return MOI.get(blmo.o, MOI.ListOfConstraintIndices{MOI.VariableIndex,MOI.LessThan{Float64}}())
+    return MOI.get(blmo.o, MOI.ListOfConstraintIndices{MOI.VariableIndex,MOI.GreaterThan{Float64}}())
```
Comment (matbesancon) [inline src/MOI_bounded_oracle.jl:78]:
```suggestion
    return MOI.get(blmo.o, MOI.ListOfConstraintIndices{MOI.VariableIndex,MOI.GreaterThan{Float64}}())
```

File: src/integer_bounds.jl [hunk 5]
```diff
@@ -79,7 +79,7 @@ end
     end
     return -1
 end =#
-
+#=
```
Comment (matbesancon) [inline src/integer_bounds.jl:82]: we can just remove the whole block instead of commenting it

Comment (dhendryc) [reply src/integer_bounds.jl:82]: Yes, we could but I wanted to leave it until I'm sure that I did not break the previous logic.

File: src/Boscia.jl (hunk 3 above)
Comment (matbesancon) [reply src/Boscia.jl:7]: > for example, we do not actually need to copy the solver right?

yes, in that case you can have the `copy` operation just return the same structure

Comment (dhendryc) [reply src/Boscia.jl:7]: I removed it. We only need this for MOI supported solvers and those have their own strong branching function.

File: src/lmo_wrapper.jl [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+"""
+Get list of variables indices. 
+If the problem has n variables, they are expected to contiguous and ordered from 1 to n.
+"""
+function get_list_of_variables end
+"""
+Get list of binary and integer variables, respectively.
+"""
+function get_binary_variables end
+function get_integer_variables end
```
Comment (matbesancon) [inline src/lmo_wrapper.jl:39]:
```suggestion

"""
Get list of binary variables.
"""
function get_binary_variables end

"""
Get list of integer variables.
"""
function get_integer_variables end 
```

File: src/lmo_wrapper.jl [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+###################################### Necessary to implement ####################################
+
+"""
+Implement `FrankWolfe.compute_extreme_point`
+
+Given a direction d solves the problem
+    min_x d^T x
+where x has to be an integer feasible point
+"""
+function compute_extreme_point end
```
Comment (matbesancon) [inline src/lmo_wrapper.jl:17]: I don't think having the empty function is needed, since we are already importing it?

File: src/lmo_wrapper.jl [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+##
```
Comment (matbesancon) [inline src/lmo_wrapper.jl:19]:
```suggestion
```

File: src/lmo_wrapper.jl [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+"""
+Get list of binary variables.
+"""
+function get_binary_variables end
+
+"""
+Get list of integer variables.
+"""
+function get_integer_variables end 
+"""
```
Comment (matbesancon) [inline src/lmo_wrapper.jl:44]: spaces between the previous function and the docstring
```julia
function f end

"""
mydoc
"""
function g end
```

File: src/MOI_bounded_oracle.jl [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+    for c_idx in consB_list
+        if c_idx.value == idx
+            return true, c_idx
+        end
+    end
+    return false, -1
+end
+
+"""
+Has variable an integer constraint?
```
Comment (matbesancon) [inline src/MOI_bounded_oracle.jl:151]:
```suggestion
Does the variable have an integer constraint?
```

File: src/cube_blmo.jl [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    for i in eachindex(v)
+        if !(blmo.bounds[i, :greaterthan].lower ≤ v[i] + 1e-6 || !(v[i] - 1e-6 ≤ blmo.bounds[i, :lessthan].upper))
+            @debug("Vertex entry: $(v[i]) Lower bound: $(blmo.bounds[i, :greaterthan].lower) Upper bound: $(blmo.bounds[i, :lessthan].upper))")
+            return false
+        end
+    end
+    return true
+end
+"""
+Has variable a binary constraint?
```
Comment (matbesancon) [inline src/cube_blmo.jl:157]:
```suggestion
Does the variable have a binary constraint?
```

File: src/MOI_bounded_oracle.jl [hunk 12]
```diff
@@ -0,0 +1,573 @@
...
+    return c_idx.value
+end
+
+"""
+Get the list of lower bounds.
+"""
+function get_lower_bound_list(blmo::MathOptBLMO) 
+    return MOI.get(blmo.o, MOI.ListOfConstraintIndices{MOI.VariableIndex,MOI.GreaterThan{Float64}}())
+end
+"""
```
Comment (matbesancon) [inline src/MOI_bounded_oracle.jl:80]:
```suggestion

"""
```

File: src/interface.jl [hunk 13]
```diff
@@ -1,9 +1,75 @@
...
+    min_fw_iterations=min_fw_iterations,
+    max_iteration_post=max_iteration_post,
+    dual_tightening=dual_tightening,
+    global_dual_tightening=global_dual_tightening,
+    bnb_callback=bnb_callback,
+    strong_convexity=strong_convexity,
+    domain_oracle=domain_oracle,
+    start_solution=start_solution,
+    fw_verbose=fw_verbose,
+    kwargs
```
Comment (matbesancon) [inline src/interface.jl:63]: I think here it did need to kwargs to also be with `kwargs...` to work properly

File: examples/approx_planted_point.jl [hunk 14]
```diff
@@ -32,8 +21,41 @@ diffi = Random.rand(Bool, n) * 0.6 .+ 0.3
...
+            MOI.add_constraint(o, xi, MOI.ZeroOne()) # or MOI.Integer()
+        end
+        lmo = FrankWolfe.MathOptLMO(o)
+
+        x, _, result = Boscia.solve(f, grad!, lmo, verbose=true)
+
+        @test x == round.(diffi)
+        @test isapprox(f(x), f(result[:raw_solution]), atol=1e-6, rtol=1e-3)
+
+        x_scip = x
```
Comment (matbesancon) [inline examples/approx_planted_point.jl:41]: variable defined and not used here?

**Pull request merged**